# Тестовое задание
на  позицию «Игровой Аналитик Данных»



Компания: разработчик mid-core игр 🎮\
Исполнитель: Сасимович Игорь (i.sasimovich@yandex.ru, Telegram: http://t.me/ihar367)


## Теоретическое задание

##### Вопрос № 1

DAU free-to-play онлайн игры 20000, количество регистраций 1500 в день, оцените месячный ретеншн.\
Объясните смысл понятия «ретеншн».

##### Ответ:

**Для оценки месячного ретеншна нам недостаточно предоставленных данных.**\
Чтобы рассчитать 30-day retention, нам нужно знать динамику количества активных пользователей определенной когорты по дням.\
Т.е. в данном случае, нам не хватает данных о числе пользователей, зарегестрированных в отсчетный день (Day 0), которые остались активными через 30 дней.

Пример расчёта месячного ретеншена, **если бы у нас были необходимые данные**:

In [1]:
# Пример:
# количество пользователей зарегистрированных 1 августа 2024 г. = 1500 юзеров
total_new_users = 1500
# количество активных пользователей вернувшихся на 30 день после первого входа в игру (из тех, что зарегистрированы 1 августа 2024 г.) = 75 юзеров
returned_users = 75

# Расчет Retention D30
retention_d30 = (returned_users / total_new_users) * 100

# Вывод результата
print(f"30-дневный ретеншн составляет {round(retention_d30, 2)}%.")

30-дневный ретеншн составляет 5.0%.


   <span style="color: #FF4500;">**Retention**</span> - одна из метрик, отражающая «удержание пользователей» игрой. Она говорит о том, продолжают ли люди пользоваться продуктом снова и снова, находят ли они его достаточно ценным (интересным/увлекательным/качественным) для себя.  Т.е. данная метрика напрямую связана с пользовательским опытом и качеством продукта. Если пользователю нравится игра или любой другой продукт, он с радостью будет в неё/него возвращаться.

Если коротко, то Retention показывает сколько людей (процент пользователей) вернулись в день Х (неделю Х, месяц Х) после первого входа в продукт (Day 0).

<span style="color: #FF4500;">**Бизнес-суть метрики Retention:**</span> Чем дольше пользователь остается в проекте, тем выше вероятность совершения платежа, его размер <span style="color: #FF7F50;"> и/или </span> больше количество просматриваемой рекламы. Соответственно высокий уровень удержания пользователей напрямую влияет на доходность и успешность продукта.

Если верить	Василию Сабирову и его книге 'Игра в цифры. Как аналитика позволяет видеоиграм жить лучше', cредне-хорошее значение удержания 1-го дня — 40%, 7-го дня — 20%, 28-го дня — 10%. Эти значения могут варьироваться в зависимости от типа игры или приложения, но они дают общее представление о том, как долго пользователи остаются активными.

##### Вопрос № 2
Месячный ARPU другой ftp он-лайн игры 46 рублей.\
Месячный ретеншн 25%.\
Стоимость привлечения 1 пользователя с помощью рекламы 61 рубль.\
Имеет ли смысл давать рекламу?\
Если нет, то при каких условиях стоит?\
Решение нужно расписать подробно, поясняя используемые формулы.

##### Ответ:

In [2]:
""" 
Чтобы понять, целесообразно ли давать рекламу, нужно сравнить стоимость привлечения одного пользователя (CPI) с тем доходом, который этот пользователь принесет игре в среднем, т.е. с метрикой LTV. 
LTV (Life Time Value) - это совокупный доход, который приносит один пользователь за все время, пока он остается активным. 
Смысл LTV: прикинуть сколько можно потратить на привлечение, т.е. должно соблюдаться следующее неравенство LTV > CPI. 

Формула расчёта: 
LTV = ARPU x Lifetime 

Lifetime - среднее время от установки до неактивности. Формула расчёта: Lifetime = 1 / Churn  ➡️ LTV = ARPU × (1 ÷ Churn) 

Churn - метрика оттока пользователей, которая оценивает долю пользователей, которые прекратили использование продукта в определенный период времени. 
Churn - обратная метрика Retention, соответственно Churn = 1 - Retention.

"""

# Исходные данные:
# ARPU (Average Revenue Per User) - средний доход с одного активного пользователя за определенный период времени (в нашем случае за месяц).(RUB). 
arpu = 46

# 30-day retention - показывает сколько людей вернулись на 30 день после первого входа в игру. 
retention_d30 = 0.25

# CPI (Cost Per Install) - метрика, которая определяет стоимость привлечения одной установки игры.(RUB).
cpi = 61

# Расчёт:

churn = 1 - retention_d30

ltv = arpu / churn

if ltv > cpi:
    print(f"LTV составляет {round(ltv,2)} RUB, что превышает CPI на {round(ltv - cpi, 2)} RUB. Возможно давать рекламу и имеет смысл.")
else:
    print(f"LTV составляет {round(ltv,2)}. Давать рекламу не имеет смысла.")

LTV составляет 61.33 RUB, что превышает CPI на 0.33 RUB. Возможно давать рекламу и имеет смысл.


<span style="color: #FF4500;">**Но с оговоркой:**</span>

Согласно приведённым расчетам, реклама потенциально может быть рентабельной, так как LTV превышает CPI. Однако следует учитывать несколько важных моментов:

* Хотя LTV (61.33 рубля) немного превышает CPI (61 рубль), это различие незначительное (0.33 рубля). Такой небольшой запас может быть недостаточным для уверенности в рентабельности рекламы, особенно если учесть возможные неопределённости, переменные затраты, качество привлечённого трафика, а также погрешности и ограничения метода расчёта LTV.

* LTV рассчитывается на основе дохода (Revenue) от пользователей и не учитывает переменные затраты, которые могут существенно повлиять на реальную прибыльность рекламы.

Таким образом, **для принятия окончательного решения о целесообразности привлечения пользователей через рекламу недостаточно данных** 😅\
Следует учитывать как доходы, так и все связанные затраты, а также возможные риски и неопределённости. 

 <span style="color: #FF4500;">*Справочно:*</span> В SaaS и других моделях бизнеса часто рекомендуется, чтобы LTV был в 3 раза больше CPI для обеспечения привлекательности бизнеса для инвесторов. В нашем случае LTV лишь немного превышает CPI, что указывает на необходимость дальнейшего улучшения финансовых показателей.


Стоит отметить, что 30-дневный ретеншен в 25% — это очень хороший показатель, указывающий на то, что игра пользуется популярностью и игроки продолжают возвращаться. Однако в игровой индустрии, где такие высокие показатели ретеншена — редкость, достижение кратного превосходства LTV над CPI может быть сложной задачей.

Кроме того, если бы у нас было больше данных о ретеншене и ARPU на 60-й и 90-й дни, мы могли бы сделать более точные выводы, но даже с текущими данными ситуация выглядит обнадеживающе.

<span style="color: #FF4500;">*Справочно:*</span> Согласно [оценке рынка мобильных игр за 2023г. от Adjust и Applovin](https://app2top.ru/analytics/otsenka-ry-nka-mobil-ny-h-igr-ot-adjust-i-applovin-upali-ne-tol-ko-zagruzki-no-i-chislo-sessij-s-retenshenom-215907.html)  средний по играм уровень удержания составил 5,3%.

<span style="color: #FF4500;">*Справочно:*</span> Если верить [этому источнику](https://www.blog.udonis.co/mobile-marketing/mobile-games/mobile-gaming-statistics), то средний 28-дневный ретеншн для верхних 25%, медианных 50% и нижних 25% мидкор игр в 2023 году составил 1.6%, 0.6% и 0.03% соответственно.

Рекомендации по улучшению:

* Важно стремиться к увеличению LTV и снижению CPI для повышения рентабельности рекламы. Например, улучшение удержания пользователей (Retention) и повышение их доходности (ARPU) может существенно увеличить LTV.
* Снижение CPI может быть достигнуто за счёт оптимизации рекламных кампаний и более точного таргетирования.


##### Вопрос № 3
Перечислите несколько способов уменьшения количества ранних отказов в free-to-play онлайн играх.

##### Ответ:

Для уменьшения количества ранних отказов в free-to-play онлайн играх (когда пользователи перестают играть вскоре после установки) / увеличения 0-day, 1-day и 7-days retention, можно использовать следующие стратегии:

* **Работа над качеством игры в целом**: 
    * Работа над интерфейсом. Он должен быть интуитивно понятным и привлекательным. Игроки не будут разбираться в сложных элементах, если на старте им не понравится дизайн и навигация.
    * **Интерактивное обучение:** простое и увлекательное обучение, чтобы новые игроки быстро освоились с основными механиками игры. Создание увлекательного и ненавязчивого туториала, который постепенно вводит игрока в механику игры и объясняет основные функции.
    * Работа над сюжетом. Старт игры с интересной завязки или интригующего события, которое увлекает игрока.
    * Баланс сложности: постепенное увеличение сложности удерживает интерес игроков, мотивируя их улучшать персонажа для преодоления трудностей. 

* Приветственные бонусы за первый вход в игру и за ежедневное возвращение. Ежедневные награды за вход.

* Еженедельные/ежедневные задания c ценными наградами. 

* Прогрессия и цели: четко сформированные ранние цели, которые легко достичь, чтобы мотивировать игроков продолжать игру.

* Специальные мероприятия, возможно, которые привязаны к событиям из реальной жизни. 

* Акции и скидки.

* Push-уведомления. Персонализированные уведомления.Таргетированные предложения.

* Непредсказуемость и случайность. Увеличение частоты неожиданных событий и сюрпризов может поддерживать интерес игроков.

* Постепенное открытие функционала (как в сериалах).

* Социализация пользователей внутри игры (например, кооперативные миссии, гильдии, чаты). Интеграция с социальными сетями.

Так же необходимо использовать аналитику для выявления моментов, когда игроки чаще всего бросают игру, и оптимизировать эти части игрового процесса. 
Важные метрики для отслеживания:

* 0-Day Retention
* 1-Day Retention
* Tutorial COR 
* Tutorial Skip Rate
* Скорость прохождения Tutorial
* Доля игроков дошедших до Уровня N
* ARPU, ARPPU, Paying share, CR в первый платеж 

* CARPU N day — накопительный ARPU за N дней (или RPI — Revenue Per Install) - позволяет определить, насколько прибыльными являются пользователи на разных стадиях их взаимодействия с игрой.
* CR в первый платеж (в окне 7 дней)
* 7-Days Retention

Особое внимание следует уделять метрикам 0-Day и 1-Day Retention. Эти показатели помогают понять, как игроки взаимодействуют с игрой на ранних этапах. Удержание 1-го дня особенно важно, поскольку оно отражает первое впечатление пользователя о проекте и его реакцию на интерфейс и функционал, и в последующем оно влияет на удержание последующих дней.

##### Вопрос № 4

Перечислите несколько способов увеличения продаж (краткосрочных и долгосрочных) конкретного игрового предмета во free-to-play игре.

##### Ответ:

Для увеличения продаж конкретного игрового предмета во free-to-play игре можно использовать как краткосрочные, так и долгосрочные стратегии. 

Краткосрочные стратегии:

1. Ограниченные по времени акции: Создание временных акций с ограниченным временем предложения, где предмет доступен по сниженной цене или с дополнительными бонусами. Это может стимулировать игроков к покупке именно сейчас, чтобы не упустить возможность.

2. Индивидуальные предложения. Использование данных о поведении игрока для создания персонализированных предложений. 

3. Внутриигровые рекламные кампании. Использование  внутриигровых рекламных баннеров, push-уведомлений и e-mail рассылок для продвижения предмета. 

4. Промо-коды и купоны: Распространение промо-кодов или купонов на скидку, которые можно использовать для покупки предмета, через социальные сети, партнерские ресурсы или внутри игры.

5. Специальные события и конкурсы: Организация игровых событий или конкурсов, в которых покупка предмета может быть вознаграждена дополнительными бонусами.

6. Комплексное предложение предмета в сочетании с временными улучшениями, такими как увеличение бонусов, опыта или игровой валюты, что может повысить его привлекательность.

7. Работа над разнообразием предметов / бандлов: Создание различных версий предмета может удовлетворить разные сегменты игроков.

8. Психологическое ценообразование (цены вида 99.99): Использование цен, заканчивающихся на .99, что может восприниматься как более выгодное предложение, стимулируя покупку. Сейчас набор "Маршал" для игроков Беларуси стоит **75 BYN**, таким образом цена с учётом психологического аспекта должна быть 74.99 или 69.99, 79.99 BYN. Такие цены работают на подсознательном уровне, делая товар визуально дешевле.



Долгосрочные стратегии:

1. Разработка уникального контента. Уникальное и ограниченное предложение. Это может быть эксклюзивный дизайн, особые характеристики предмета или редкость.

2. Регулярное обновление контента: Поддержание интереса и привлечение как старых, так и новых игроков.

3. Внедрение в сюжет и геймплей: Интеграция предмета в основной сюжет или геймплей, чтобы игроки почувствовали его важность и необходимость для достижения целей или выполнения квестов.

4. Лояльность и программы вознаграждений: Введение программ лояльности, где игроки получают бонусы за регулярные покупки.

5. Социальные и кооперативные элементы:  Введение возможностей для игроков демонстрировать свои приобретения другим игрокам, что стимулирует желание купить предмет.


Для повышения эффективности этих стратегий важно постоянно анализировать данные по продажам и поведению игроков. Моменты, которые стоит учитывать:

* Структура покупок: Анализ покупок по уровням, сегментам и времени.
* Потребительская корзина: Анализ товаров, часто покупаемых вместе. Динамика корзины.
* Взаимодействие товаров: Определение, как продажа одного товара влияет на другой.
* Распределение по уровням и времени: Изучение, на каких этапах игры и в какое время продажи предмета наиболее эффективны.
* Стабильность спроса: Оценка долгосрочного спроса на предмет и определение пиковых периодов продаж.

Так же могут быть полезными: ABC-анализ и XYZ-анализ.
Это инструменты, которые используются для сегментации товаров, чтобы оптимизировать управление запасами, маркетингом и продажами.
ABC-анализ позволяет классифицировать игровые предметы по важности на основе дохода, который они приносят. 
XYZ-анализ дополняет ABC-анализ, классифицируя предметы по стабильности спроса.
И их комбинированная версия - ABC-XYZ.
Используя ABC-XYZ анализ,  можно более точно направлять свои усилия на те игровые предметы, которые принесут максимальную отдачу, при этом оптимизируя расходы на продвижение и разработку контента.

## Практические задание

##### Вопрос № 1

Вычислите примерные доходности основных ПвП и ПвЕ режимов (в монетах в час).\
Рассчитайте минимальное время получения пушки “Палач 88 мм” (в часах).\
Все расчеты снабдите подробными формулами и поясните логику рассуждений.

##### Ответ:
В Crossout Mobile есть две основные валюты и аж 8 ресурсов:

* Талеры — серебряные жетоны, необходимые для производства модулей для бронемобиля. Талеры выдаются в качестве награды за матчи и за выполнение заданий. Их также можно приобрести за монеты.
* Монеты — золотые жетоны, премиальная валюта / хард валюта. Нужны для приобретения модулей и расходников на рынке, производства эпических и более редких деталей, ускорения таймеров производства.

Ресурсы, которые нужны для крафта деталей, получаются в качестве награды за бой или за выполнение задания, а так же в качестве бонусов: металлолом, провода, медь, аккумуляторы, пластик, микросхемы, гильзы, урановая руда. Все они различаются по редкости — чем более редкий ресурс, тем он дороже и сложнее его достать.

Теперь перейдём к расчету доходности основных PVP и PVE режимов (в монетах в час).
Монеты не являются наградой за непосредственно PVP и PVE бои. Не покупая их за реальные деньги, без покупки боевого пропуска и премиум подписки, их можно заработать в небольших количествах за выполнение некоторых заданий, и как награду за продвижение по пути выжившего (максимум 300 монет в месяц), а также торговлю на рынке. Также, они даются, как компенсация за получение дубликата запчасти из тайника.

**Поскольку монеты не выдаются напрямую за PVP или PVE бои, то нам не остаётся ничего другого, как оценить их доходность (в монетах в час) косвенно через ресурсы. В этом нам поможет игровой рынок с динамическим ценообразованием.**\
*Данные актуальны на 29 августа 2024 года.*

| Ресурсы              | Партия | Актуальная цена на рынке на 29.08.2024 | Курс (монет за 1 ед. ресурса) |
|----------------------|--------|---------------------------------------|-------------------------------|
| Ресурс "Аккумуляторы" | 5000   | 98                                    | 98 / 5000 = 0.0196            |
| Ресурс "Пластик"      | 5000   | 87                                    | 87 / 5000 = 0.0174            |
| Ресурс "Провода"      | 10000  | 12                                    | 12 / 10000 = 0.0012           |
| Ресурс "Медь"         | 10000  | 80                                    | 80 / 10000 = 0.0080           |
| Ресурс "Металлолом"   | 50000  | 40                                    | 40 / 50000 = 0.0008           |

Как видно из таблицы, наиболее дорогим/высокодоходным ресурсом являются "Аккумуляторы".

Рассчитывать будем по следующей методике:
1.	Сначала рассчитаем возможное количество игр за час с учётом времени необходимого на поиск матчей (пауза 1,5 мин.). 
2.	Посчитаем общее количество ресурсов, которых игрок может потенциально заработать за 1 час.
3.	Переведём ресурсы в монеты по рыночному курсу, установленному на момент выполнения расчёта.

Для расчёта нам понадобятся следующие данные, полученные эмпирическим путём, так что поиграем и зафиксируем результаты:

| Режим игры                     | Среднее вознаграждение (ед.)| Длительность игры (мин.) | Длительность одной игры с паузой (мин.) |
|--------------------------------|-----------------------------|--------------------------|-----------------------------------------|
| **PVE**                        |                             |                          |                                         |
| PVE за ресурс "Металлолом"     | 176                         | 10                       | 11,5                                    |
| PVE за ресурс "Медь"           | 274                         | 8                        | 9,5                                     |
| PVE за ресурс "Пластик"        | 57                          | 7,5                      | 9                                       |
| **PVP**                        |                             |                          |                                         |
| PVP за ресурс "Металлолом"     | 70                          | 3                        | 4,5                                     |
| PVP за ресурс "Провода"        | 130                         | 3                        | 4,5                                     |
| PVP за ресурс "Медь"           | 115                         | 3                        | 4,5                                     |
| PVP за ресурс "Пластик"        | 84                          | 3                        | 4,5                                     | 
| PVP за ресурс "Аккумуляторы"   | 96                          | 3                        | 4,5                                     | 

⚠️ На самом деле, величина среднего вознаграждения может значительно отличаться. Размер вознаграждения за бой очень зависит от следующих факторов: режима игры, уровня игрока, мощности его боевой техники, исхода боя (победа/проигрыш/ничья), эффективности игрока в бою.\
Наиграв суммарно более 200 игр и собрав машину мощностью 6800 ОМ, одержать победу в боях за "Пластик" и "Аккумуляторы" стало не совсем простой задачей 😅 Так что, данные показатели характеризуют скорее мою игру.

**Для примера рассчитаем доходность режима  PVE за ресурс "Металлолом":**

1. Учитывая 1,5 минуты паузы между играми, длительность одной игры составит 10 + 1,5 = 11,5 минут.
2. Рассчитаем количество игр за час, которое может сыграть игрок: 60 / 11,5 ≈ 5,22 игр (округлим до 5).
3. За 5 игр игрок заработает металлолома: 5 * 176 = 880 ед.
4. Стоимость металлолома на рынке составляет 40 монет за 50 000 ед. Используя эту информацию, переведём ресурс в монеты: 40 / 50 000 * 880 = 0,704 монеты.

Подобным образом рассчитаем доходность остальных режимов и получим следующую таблицу:

Результаты расчётов: 
| Режим игры                            | Доходность в час (монеты) |
|---------------------------------------|---------------------------|
| **PVE**                               |                           |
| PVE за ресурс "Металлолом"            | 0,704                     |
| PVE за ресурс "Медь"                  | 13,152                    |
| PVE за ресурс "Пластик"               | 16,386                    |
| **PVP**                               |                           |
| PVP за ресурс "Металлолом"            | 0,728                     |
| PVP за ресурс "Провода"               | 2,028                     |
| PVP за ресурс "Медь"                  | 11,96                     |
| PVP за ресурс "Пластик"               | 19,0032                   |
| PVP за ресурс "Аккумуляторы"          | 24,4416                   |

**При текущем внутриигровом уровне цен на ресурсы, наиболее высокодоходным режимом, с доходностью 24,44 монеты в час, является PVP за ресурс "Аккумуляторы". Учитывая динамическое ценообразование, ситуация  может измениться.**\
**Однако, данный расчёт весьма упрощён и не учитывает такие факторы, как эффективность игрока, мощность его боевой техники, исход боя, которые влияют на размер получаемых ресурсов за бой.**
**Также не учитываются различного типа бонусы и награды за выполнение заданий.**\
*Справочно:* у игрока есть возможность повысить доходность боя в 2 раза, оформив премиум подписку, которая даёт Х2 на вознаграждения за бои и другие плюшки.

--- 

**В Crossout Mobile есть 2 способа получения эпической пушки "Палач 88 мм":**
1) крафт на основе чертежа, полученного в ходе выполнения заданий 'Дорога славы' и достижения игроком 24 уровня во фракции "Мусорщики";
2) покупка данной пушки на рынке за монеты. На 29 августа 2024 года, 12:10, актуальная цена пушки "Палач 88 мм" составляет 221 монету.

Так вот, затраты на крафт 1 ед. детали: пушка “Палач 88 мм” составляют:

| Компоненты            | Количество |
|-----------------------|------------|
| Монеты                | 252        |
| Ресурс "Аккумуляторы" | 4 700      |
| Ресурс "Пластик"      | 4 700      |
| Ресурс "Провода"      | 7 975      |
| Ресурс "Медь"         | 7 975      |
| Ресурс "Металлолом"   | 8 885      |
| Талеры               | 21 210     |


Как выяснилось в ходе изучения данного вопроса: для крафта эпической пушки "Палач 88 мм" тоже необходима хард валюта - монеты, да ещё и больше, аж 252 монеты. А если перевести все затраты в монеты, то себестоимость крафта пушки составит 506,32 монет и 21 210 талеров. Талеры не торгуются на рынке, их можно купить за монеты в магазине по цене 10 000 талеров за 100 монет, т.е., если выразить талеры в монетах, то получим 212,1 монет.\
**Итоговая себестоимость крафта эпической пушки "Палач 88 мм" составит около 718,5 монет.**\
**Очевидно, при текущих условиях на рынке крафт 'Палача' крайне невыгоден и значительно уступает покупке..**

Чтобы рассчитать минимальное время получения пушки "Палач 88 мм", разделим рыночную стоимость пушки на максимальную доходность в час:

Т.е. время получения пушки = стоимость ее на рынке / максимальную доходность в час = 221 / 24,4416 = 9,04 часа

Время получения данной детали через крафт составит 29,39 часов.


Ответ:
Минимальное время получения пушки "Палач 88 мм" при текущих условиях составляет примерно 9,04 часа, при условии, что игрок зарабатывает на максимальной доходности, но это может изменяться в зависимости от эффективности игры игрока и его навыков, изменений в игре и на её рынке. Крафт данной пушки займет около 30 часа активной игры (и это не учитывая 8 часов ожидания), что делает этот путь получения пушки весьма затратным по времени.
Сравнение рыночной стоимости и затрат на крафт пушки показывает, что покупка на рынке сэкономит игроку до 21 часа активной игры, **а покупка хард валюты вовсе все 30** 😂

Как я уже писал в тестовом на позицию "Операционный менеджер", на мой взгляд, рынок в «Crossout mobile» делает внутриигровую экономику более  интересной и сбалансированной, а игру более сложной в хорошем смысле слова (не примитивной). В игре появляется экономический аспект. Кроме того, он обеспечивает баланс между игровыми усилиями и возможностью зарабатывать хард валюту без необходимости доната, что делает игру более доступной и интересной для всех категорий игроков.

Постоянный мониторинг рыночных цен и стоимости ресурсов поможет игрокам своевременно адаптировать свои стратегии, чтобы минимизировать затраты и максимизировать эффективность.
Таким образом, игроки могут использовать информацию о доходности и рыночных ценах для оптимизации своих стратегий получения редких предметов. Баланс между затратами времени и ресурсов позволяет сделать осознанный выбор между крафтом и покупкой, что добавляет игре стратегическую глубину.

При этом разработчикам  важно следить за соотношением затрат на крафт и покупку, поскольку разница в стоимости может значительно влиять на баланс игровой экономики. Если крафт становится слишком дорогим по сравнению с покупкой, это может снижать его привлекательность и вынуждать игроков больше полагаться на рынок. Игроки, осознав невыгодность крафта, могут полностью отказаться от этой механики, что может сделать игру менее интересной и разнообразной. Если игроки видят, что их усилия по добыче ресурсов и крафту не окупаются, они могут потерять интерес к игре, особенно если чувствуют, что покупка предметов за реальные деньги становится единственным реальным способом получения нужных предметов.

##### Вопрос № 2

Предположим, что ресурс “Провода” стоит дороже ресурса “Аккумуляторы”.\
Будет ли это проблемой для игры?\
Какими способами можно увеличить цену “Аккумуляторов”?\
Опишите свою логику в виде ТЗ.

##### Ответ в виде ТЗ:

**Техническое задание:** Анализ и корректировка цен на ресурсы в Crossout Mobile

*Цель:* Оценить, является ли ситуация, при которой ресурс “Провода” стоит дороже ресурса “Аккумуляторы”, проблемой для игры, и предложить способы увеличения цены “Аккумуляторов”.

**1. Описание ресурсов и их потребления**

<span style="color: #4682B4;">**"Провода"**</span>
- Доступность: При достижении боевым транспортом игрока 3300 ОМ.
- Использование: Крафт деталей редкой (синей) редкости.

<span style="color: #6A5ACD;">**"Аккумуляторы"**</span>
- Доступность: При достижении 5200 ОМ.
- Использование: Крафт деталей эпической (фиолетовой) редкости.


В Crossout Mobile детали различаются по редкости и мощи:
 - обычные (помечены серым, самые слабые), 
 - <span style="color: #4682B4;">редкие</span> (помечены синим), 
 - <span style="color: #6A5ACD;">эпические</span> (помечены фиолетовым), 
 - легендарные (помечены желтым, для новичка практически недоступны) 
 - реликтовые (помечены оранжевым, получить их крайне сложно)

Для крафта вещей уровня выше обычных используется 2 вида ресурсов, например, для крафта редкой вещи - "Провода" и "Медь".
Стоит отметить, что  перечень компонентов для крафта деталей высшего уровня, включает в себя деталь, а то и несколько деталей предыдущего уровня. Так, например, для крафта эпической башенной пушки "ЗИС-33 Верзила" помимо ресурсов "Пластик" и "Аккумуляторы",  понадобятся следующие редкие детали: 2 ед. пушки "Малыш 6Ф" и 2 ед. дробовика "Очаг".

Факторы, определяющие ценность ресурсов, как в игре, так и в реальной жизни:
1. Редкость
2. Доступность
3. Полезность и функциональность
4. Спрос и предложение
5. Заменяемость

**2. Оценка возможных проблемы**

Неправильное соотношение цен (“Провода” стоят дороже “Аккумуляторов”) может привести к следующим проблемам:

* Нарушению игрового баланса и экономики: 
   * Нарушение баланса между редкостью ресурсов и их ценой может привести к экономическим проблемам в игре, таким как избыток или дефицит ресурсов, что нарушит игровой процесс и экономику. 
   * Дисбаланс в ценах может снизить ценность “Аккумуляторов” и привести к уменьшению их значимости для крафта эпических предметов.

* Нарушению баланса сложности:
   * Данная ситуация мотивирует более опытных и сильных игроков участвовать в PVP/PVE режимах по добыче "Проводов", в которых играют игроки послабее, что может негативно отразиться на мотивации последних играть в игру. 
   * Данная ситуация может привести к тому, что создание эпических предметов станет менее сложным и затратным, а редких наоборот, что нарушает баланс между различными уровнями предметов и в целом логику крафта, что может снижать интерес к игре. 

* ⚠️ Снижение объёма внутриигровых продаж 
   * Легкодоступность эпических предметов может снизить ценность уникальных наборов и хард валюты, что уменьшит стимул к их покупке. Как результат, снижение доходов от внутриигровых покупок, что негативно скажется на финансовой модели игры.


**3. Способы увеличения цены “Аккумуляторов”**

1. **Создание дефицита ресурса путём увеличения его потребления или сокращения его доступности:**
   - Описание: Ограничить количество “Аккумуляторов”, которые можно получить из ежедневных заданий, контейнеров или торговли.
   - Пример: Снижение частоты получения “Аккумуляторов” в наградах или контейнерах.
  
   **Изменение затрат на крафт:**
   - Описание: Увеличить количество “Аккумуляторов”, необходимых для создания эпических предметов.
   - Пример: Увеличение количества “Аккумуляторов” для крафта эпической детали на 20-30%.

2. **Изменение базовой/актуальной цены:**
   - Описание: Увеличить базовую цену “Аккумуляторов” на рынке.
   - Пример: Увеличение базовой цены с 20 единиц валюты до 30-40 единиц.

3. **Регулировка динамического ценообразования:**
   - Описание:  Для предотвращения резких колебаний цен и возможных дисбалансов, необходимо установить механизмы регулирования, которые будут ограничивать чрезмерное падение или рост цен. Это обеспечит стабильность рынка и предотвратит ситуации, в которых игроки могут злоупотреблять ценовыми изменениями.
   - Пример: Использование алгоритмов для автоматического изменения цен на ресурсы в зависимости от рыночного спроса, с введением верхних и нижних пределов ценовых колебаний. Например, если спрос на "Аккумуляторы" резко падает, система может ограничить снижение цены до определённого уровня, чтобы избежать обесценивания ресурса.

4. **Игровые события и акции:**
   - Описание: Проведение специальных событий и акций, в которых "Аккумуляторы" можно будет поменять на уникальный предмет.
   - Пример: Проведение временных акций с обменом “Аккумуляторов” на режим "Без рекламы".

5. **Введение уникальных предметов:**
   - Описание: Создать новые предметы или улучшения, требующие большого количества “Аккумуляторов”.
   - Пример: Новые эпические предметы, которые требуют увеличенное количество “Аккумуляторов” для крафта.

**4. Заключение**: 
Для поддержания баланса в игре важно, чтобы ценность ресурсов была адекватно отражена в их цене. Если "Провода" остаются дороже "Аккумуляторов", это может нарушить естественное распределение ресурсов и прогрессии в игре. Низкая цена на "Аккумуляторы" уменьшит сложность крафта эпических предметов, что повлияет на экономику игры и её долгосрочную привлекательность для игроков. Поэтому необходимо провести корректировки, направленные на увеличение цены "Аккумуляторов" для восстановления игрового баланса и поддержания интереса к процессу создания высокоуровневых предметов.

Рекомендуется регулярно мониторить экономику игры, отслеживая изменения в спросе и предложении на ресурсы, и вносить соответствующие изменения, чтобы поддерживать стабильный и увлекательный игровой процесс. Эти меры помогут избежать дисбаланса и поддержат активное участие игроков, сохраняя их интерес к игре и её экономической системе на долгосрочной основе.

В Crossout Mobile уже применяются все эти методы, и это делает игру такой увлекательной, создавая гармоничную и сбалансированную игровую экосистему.

Поддержание справедливой и динамичной экономики внутри игры не только укрепляет интерес текущих игроков, но и привлекает новых, создавая сообщество, которое ценит баланс и продуманность каждой детали. 

##### Вопрос № 3

Есть реляционная база данных следующей структуры:
* Таблица users содержит данные о регистрации пользователей в игре с разбивкой по датам\
    CREATE TABLE IF NOT EXISTS users (\
    uid bigint NOT NULL primary key, // идентификатор пользователя\
    reg_time timestamp // дата и время регистрации пользователя );

* Таблица dailyusers содержит данные , в какие дни пользователь играл\
    CREATE TABLE IF NOT EXISTS dailyusers (\
    uid bigint NOT NULL, // идентификатор пользователя\
    day date NOT NULL, // дата, когда он играл\
    CONSTRAINT\
    dailyusers_pkey PRIMARY KEY (uid, day) );\
    
Требуется написать sql-запросы, которые вычисляют следующие метрики:\
   * A. Число активных пользователей в день.\
   * B. Число зарегистрированных пользователей в день.\
   * C. Однодневный классический ретеншн (классическое удержание).\
   * D. Однодневный роллинг ретеншн (скользящее удержание).

Результат каждого SQL запроса должен содержать 2 колонки:
- дата
- значение метрики.

Желательно использовать синтаксис PostgreSQL.

##### Ответ:

In [3]:
# A. Число активных пользователей в день / DAU

# Очень хочется, но нет необходимости использовать DISTINCT,т.к. в таблице `dailyusers` гарантирована уникальность записей по комбинации uid и day. Данная комбинация является составным первичным ключом.

query_a = """
SELECT day, COUNT(uid) AS dau
  FROM dailyusers
 GROUP BY day
"""
# Код проверен и работает корректно на PostgreSQL 16.4.

In [4]:
# B. Число зарегистрированных пользователей в день.

#  В таблице users, согласно предоставленной структуре, уникальность uid гарантирована, поскольку uid является первичным ключом.

query_b = """
SELECT reg_time::date AS reg_date, COUNT(uid) AS reg_users
  FROM users
 GROUP BY reg_date
"""
# Код проверен и работает корректно на PostgreSQL 16.4.

In [5]:
# C. Однодневный классический ретеншн (классическое удержание).

# 1. CTE user_activity:
# - Соединяем таблицы users и dailyusers по идентификатору пользователя (uid).
# - Рассчитываем количество дней между датой регистрации (reg_date) и датой визита (visit_date).
# - Поскольку в таблице `dailyusers` гарантирована уникальность записей по комбинации uid и day, дополнительная агрегация для устранения дубликатов не требуется.
# - В результате подзапроса получаем строки с user_id, reg_date, visit_date, и days_since_reg (количество дней от регистрации).
# 2. Основной запрос. Рассчитываем однодневное удержание (RR1) в рамках каждой когорты по дате регистрации.:
# - Используем условный оператор CASE для подсчета количества пользователей, которые вернулись на 1-й день после регистрации.
# - Делим это количество на общее число уникальных пользователей в рамках когорты и умножаем на 100, чтобы получить процент.


# 1ая версия запроса с использованием CTE ❌
query_c1 = """
WITH user_activity AS (
     SELECT u.uid AS user_id, 
            u.reg_time::date AS reg_date, 
            du.day AS visit_date, 
            (du.day - u.reg_time::date) AS days_since_reg 
      FROM users u
      LEFT JOIN dailyusers du 
        ON u.uid = du.uid
)

SELECT reg_date,
       SUM(CASE WHEN days_since_reg = 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(user_id) AS rr1
  FROM user_activity
 GROUP BY reg_date
 ORDER BY reg_date;
"""
# Изменив значение WHEN days_since_reg = 1 в основном запросе на WHEN days_since_reg = 7, мы получим 7 day Retention.
# В запросе имеются дополнительные вычисления, что может быть не эффективным при работе с большим объемом данных.
# Попробуем его доработать и оптимизировать, и получим...


# 2ой вариант запроса с использованием CTE ❌
query_c2 = """
WITH user_activity AS (
     SELECT u.uid AS user_id, 
            u.reg_time::date AS reg_date, 
            du.day AS visit_date
      FROM users u
      LEFT JOIN  dailyusers du 
        ON u.uid = du.uid
)

SELECT reg_date,
       COUNT(CASE WHEN visit_date = reg_date + INTERVAL '1 day' THEN 1 END) * 100.0 / COUNT(user_id) AS rr1
  FROM user_activity
 GROUP BY reg_date
 ORDER BY reg_date;
"""
# В этом варианте  убраны лишние вычисления, что должно сделать запрос более эффективным. Однако, возможно использование CTE избыточно.
# Попробуем избавиться от CTE. В результате получим...


# 3ий вариант запроса без CTE ✅
query_c3 = """
SELECT u.reg_time::date AS reg_date,
       COUNT(CASE WHEN du.day = u.reg_time::date + INTERVAL '1 day' THEN 1 END) * 100.0 / COUNT(*) AS rr1
  FROM users u 
  LEFT JOIN  dailyusers du 
    ON u.uid = du.uid
 GROUP BY u.reg_time::date
 ORDER BY reg_date;
"""
# 3ий вариант запроса - наилучший вариант по производительности и простоте, поскольку он напрямую решает задачу без использования CTE и лишних вычислений.

# Код проверен и работает корректно на PostgreSQL 16.4.

In [6]:
# D. Однодневный роллинг ретеншн (скользящее удержание).

# Rolling Retention показывает, какой процент новых пользователей вернулись в продукт в конкретный день с момента прихода или любой день после него.

# Слегка изменим предыдущий запрос, а именно:
# 1. Исходя из определения Rolling Retention, нам нужно найти последнюю дату активности пользователя.
# Поэтому в CTE user_activity оставим только последнюю/max дату входа для каждого пользователя.
# 2. В основном запросе изменим условие на WHEN days_since_reg >= 1, чтобы учесть тех пользователей, которые вернулись на 1-й день или позже.

# 1ая версия запроса с использованием CTE ❌
query_d1 = """
WITH user_activity AS (
     SELECT u.uid AS user_id, 
            u.reg_time::date AS reg_date, 
            MAX(du.day) AS last_return_date,  
            (MAX(du.day) - u.reg_time::date) AS days_since_reg
      FROM users u
      LEFT JOIN dailyusers du 
        ON u.uid = du.uid
     GROUP BY u.uid, u.reg_time::date
)

SELECT reg_date,
       SUM(CASE WHEN days_since_reg >= 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(user_id) AS rolling_rr1
  FROM user_activity
 GROUP BY reg_date
 ORDER BY reg_date;
"""
# В запросе имеются дополнительные вычисления, что может быть не эффективным при работе с большим объемом данных.
# Попробуем его доработать и оптимизировать, и получим...


# 2ой вариант запроса с использованием CTE ✅
query_d2 = """
WITH user_activity AS (
     SELECT u.uid AS user_id, 
            u.reg_time::date AS reg_date, 
            MAX(du.day) AS last_return_date  
	   FROM users u
       LEFT JOIN dailyusers du 
         ON u.uid = du.uid
      GROUP BY u.uid, u.reg_time::date
)

SELECT reg_date,
       COUNT(CASE WHEN last_return_date  >= reg_date + INTERVAL '1 day' THEN 1 END) * 100.0 / COUNT(user_id) AS rolling_rr1
  FROM user_activity
 GROUP BY reg_date
 ORDER BY reg_date;
"""
# В этом варианте  убраны лишние вычисления, что должно сделать запрос более эффективным. Однако, возможно использование CTE избыточно.
# Попробуем избавиться от CTE. В результате получим...

# 3ая версия без CTE  ❌
query_d3 = """
SELECT u.reg_time::date AS reg_date,
       COUNT(DISTINCT du.uid) * 100.0 / COUNT(DISTINCT u.uid) AS rolling_rr1
  FROM users u
  LEFT JOIN  dailyusers du 
    ON u.uid = du.uid AND du.day > u.reg_time::date 
 GROUP BY u.reg_time::date
 ORDER BY reg_date;

"""
# 3яя версия кажется более простой, но отрабатывает в двое дольше первых двух, вероятно, из-за дополнительных вычислений в LEFT JOIN и COUNT(DISTINCT ...).
# Таким образом наиболее эффективной версией будет версия номер 2.

# Код проверен и работает корректно на PostgreSQL 16.4.

##### 4. Решите задачу.
Игроку выпало 3 одинаковых ящика. Открыв любой ящик игрок с шансом 30% может получить один уникальный предмет.\
Рассчитайте каковы шансы у игрока получить 0, 1, 2 или 3 уникальных предмета, открыв все 3 ящика.\
Обязательно объясните решение данной задачи.

##### Ответ:

Для решения данной задачи будем использовать формулу Бернулли (формулу вероятности k успехов в серии из n независимых испытаний).\
Эта формула имеет смысл только для бинарных событий. Как раз наш случай. 

![A](https://konspekta.net/megaobuchalkaru/imgbaza/baza10/3730778290776.files/image062.gif)

где:

- C(n, k) — число сочетаний (биномиальный коэффициент)
- p  — вероятность успеха в одном испытании,
- q  — вероятность неудачи в одном испытании. q = (1 - p),
- n  — общее количество испытаний,
- k  — количество успехов/ интересующих нас событий.


In [7]:
# Исходные данные
p = 0.3
q = 1 - p
n = 3
k0 = 0
k1 = 1
k2 = 2
k3 = 3

In [8]:
#1 Рассчитаем  шансы  игрока получить 0 уникальных предметов.

# Для этого согласно формуле Бернулли сначала рассчитаем биномиальный коэффициент C(3, 0). 
# Формула расчёта: C(n, k) = n! / (k! * (n - k)!)

Cnk0 = (1 * 2 * 3) / (1 * (1 * 2 * 3))
Cnk0

# Это значение имеет простой смысл: существует только один способ выбрать 0 объектов из 3 доступных, и это "выбрать ничего".

# можно проверить себя
# import math 
# Cnk0 = math.comb(n, k0) 

1.0

In [9]:
# Теперь у нас имеются все необходимые для расчёта данные. Подставим значения.
probability_0 = Cnk0 * (p ** k0) * ((1 - p) ** (n - k0))
probability_0

print(f"Вероятность получить {k0} уникальных предметов составляет {round(probability_0,3)}.") 

Вероятность получить 0 уникальных предметов составляет 0.343.


In [10]:
#2 Теперь поступим по аналогии и рассчитаем  шансы  игрока получить 1 уникальный предмет.

# Снова рассчитаем биномиальный коэффициент C(3, 1)
Cnk1 = (1 * 2 * 3) / (1 * (1 * 2))
Cnk1

3.0

In [11]:
probability_1 = Cnk1 * (p ** k1) * ((1 - p) ** (n - k1))
probability_1

print(f"Вероятность получить {k1} уникальных предметов составляет {round(probability_1,3)}.") 

Вероятность получить 1 уникальных предметов составляет 0.441.


In [12]:
#3. По аналогии  рассчитаем  шансы  игрока получить 2 уникальных предмета.
# Снова рассчитаем биномиальный коэффициент C(3, 2)
Cnk2 = (1 * 2 * 3) / ((1*2) * (1))
Cnk2

3.0

In [13]:
probability_2 = Cnk2 * (p ** k2) * ((1 - p) ** (n - k2))
probability_2

print(f"Вероятность получить {k2} уникальных предметов составляет {round(probability_2,3)}.") 

Вероятность получить 2 уникальных предметов составляет 0.189.


In [14]:
#3.И наконец  рассчитаем  шансы  игрока получить 3 уникальных предмета.
# Снова рассчитаем биномиальный коэффициент C(3, 3)
Cnk3 = (1 * 2 * 3) / ((1*2*3) * (1))
Cnk3

1.0

In [15]:
probability_3 = Cnk3 * (p ** k3) * ((1 - p) ** (n - k3))
probability_3

print(f"Вероятность получить {k3} уникальных предметов составляет {round(probability_3,3)}.") 

Вероятность получить 3 уникальных предметов составляет 0.027.


Ответ:

Вероятность получить 0 уникальных предметов составляет 0.343.\
Вероятность получить 1 уникальных предметов составляет 0.441.\
Вероятность получить 2 уникальных предметов составляет 0.189.\
Вероятность получить 3 уникальных предметов составляет 0.027.

Таким образом, наиболее вероятным исходом является получение 1 уникального предмета.